File: aoc15/day_07/day_07.ml
module Wires = Map.Make (String)

type op = IN | AND | OR | LSHIFT | RSHIFT | NOT
type rule = { in1 : string; in2 : string; out : string; op : op };;

let parse line =
  let split = Str.bounded_split (Str.regexp " -> ") line 2 in
  let left = List.nth split 0 in
  let out = List.nth split 1 in
  let re = Str.regexp {|\([a-z0-9]*\) ?\([A-Z]+\) \([a-z0-9]+\)|} in
  let m = Str.string_match re left 0 in
  let in1 = if m then Str.matched_group 1 left else left in
  let in2 = if m then Str.matched_group 3 left else "" in
  let op =
    if m then
      match Str.matched_group 2 left with
      | "AND" -> AND
      | "OR" -> OR
      | "LSHIFT" -> LSHIFT
      | "RSHIFT" -> RSHIFT
      | "NOT" -> NOT
      | _ -> IN
    else IN
  in
  { in1; in2; out; op }
in

let rules = Input.parse_lines parse "input" in

let valid wires i =
  i = "" || int_of_string_opt i != None || Wires.mem i !wires
in
let next_rule wires r =
  valid wires r.in1 && valid wires r.in2 && not (Wires.mem r.out !wires)
in
let get wires x =
  match int_of_string_opt x with Some n -> n | None -> Wires.find x !wires
in
let apply wires rule =
  match rule.op with
  | IN -> get wires rule.in1
  | AND -> get wires rule.in1 land get wires rule.in2
  | OR -> get wires rule.in1 lor get wires rule.in2
  | LSHIFT -> get wires rule.in1 lsl get wires rule.in2
  | RSHIFT -> get wires rule.in1 lsr get wires rule.in2
  | NOT -> lnot (get wires rule.in2)
in

let compute init_wires =
  let wires = ref init_wires in
  let quit = ref false in
  while not !quit do
    let rule = List.find (next_rule wires) rules in
    let out = apply wires rule in
    wires := Wires.add rule.out out !wires;
    if Wires.mem "a" !wires then quit := true
  done;
  Wires.find "a" !wires
in

let first = compute Wires.empty in
Printf.printf "First: %d\n" first;

let second = compute (Wires.singleton "b" first) in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_07/day_07.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func value(w string, wires map[string]chan uint16) uint16 {
	v, err := strconv.Atoi(w)
	if err == nil {
		return uint16(v)
	} else {
		v := <-wires[w]
		return v
	}
}

func eval(name, rule string, wires map[string]chan uint16) {
	var val uint16
	split := strings.Fields(rule)
	if len(split) == 1 {
		val = value(split[0], wires)
	} else if split[0] == "NOT" {
		val = ^value(split[1], wires)
	} else {
		x := value(split[0], wires)
		y := value(split[2], wires)
		switch split[1] {
		case "AND":
			val = x & y
		case "OR":
			val = x | y
		case "LSHIFT":
			val = x << y
		case "RSHIFT":
			val = x >> y
		}
	}
	for {
		wires[name] <- val
	}
}

func run(rules map[string]string, wires map[string]chan uint16) uint16 {
	for name, rule := range rules {
		go eval(name, rule, wires)
	}
	res := <-wires["a"]
	return res
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	rules := make(map[string]string)
	wires := make(map[string]chan uint16)
	for scanner.Scan() {
		rule := strings.Split(scanner.Text(), " -> ")
		rules[rule[1]] = rule[0]
		wires[rule[1]] = make(chan uint16)
	}

	first := run(rules, wires)
	fmt.Println("First:", first)

	time.Sleep(100 * time.Millisecond)

	rules["b"] = strconv.Itoa(int(first))
	for w := range wires {
		wires[w] = make(chan uint16)
	}
	second := run(rules, wires)
	fmt.Println("Second:", second)
}
